File: frontend/src/components/benchmark/content/quick_sort/implementations/quick_sort.go
package main

import (
    "math/rand"
)

type QuickSort struct{}

func (qs *QuickSort) partition(arr []int, low, high int) int {
    pivot := arr[high]
    i := low - 1

    for j := low; j < high; j++ {
        if arr[j] <= pivot {
            i++
            arr[i], arr[j] = arr[j], arr[i]
        }
    }
    arr[i+1], arr[high] = arr[high], arr[i+1]
    return i + 1
}

func (qs *QuickSort) partitionRandom(arr []int, low, high int) int {
    randomIdx := rand.Intn(high-low+1) + low
    arr[randomIdx], arr[high] = arr[high], arr[randomIdx]
    return qs.partition(arr, low, high)
}

func (qs *QuickSort) quickSortRecursive(arr []int, low, high int) {
    if low < high {
        pi := qs.partition(arr, low, high)
        qs.quickSortRecursive(arr, low, pi-1)
        qs.quickSortRecursive(arr, pi+1, high)
    }
}

func (qs *QuickSort) quickSortRandomRecursive(arr []int, low, high int) {
    if low < high {
        pi := qs.partitionRandom(arr, low, high)
        qs.quickSortRandomRecursive(arr, low, pi-1)
        qs.quickSortRandomRecursive(arr, pi+1, high)
    }
}

func (qs *QuickSort) Sort(arr []int) {
    qs.quickSortRecursive(arr, 0, len(arr)-1)
}

func (qs *QuickSort) SortRandom(arr []int) {
    qs.quickSortRandomRecursive(arr, 0, len(arr)-1)
}

func (qs *QuickSort) SortIterative(arr []int) {
    stack := [][2]int{{0, len(arr) - 1}}

    for len(stack) > 0 {
        top := stack[len(stack)-1]
        stack = stack[:len(stack)-1]
        low, high := top[0], top[1]

        if low < high {
            pi := qs.partition(arr, low, high)
            stack = append(stack, [2]int{low, pi - 1})
            stack = append(stack, [2]int{pi + 1, high})
        }
    }
}

func (qs *QuickSort) threeWayPartition(arr []int, low, high int) {
    if low >= high {
        return
    }

    lt, gt := low, high
    pivot := arr[low]
    i := low

    for i <= gt {
        if arr[i] < pivot {
            arr[lt], arr[i] = arr[i], arr[lt]
            lt++
            i++
        } else if arr[i] > pivot {
            arr[i], arr[gt] = arr[gt], arr[i]
            gt--
        } else {
            i++
        }
    }

    qs.threeWayPartition(arr, low, lt-1)
    qs.threeWayPartition(arr, gt+1, high)
}

func (qs *QuickSort) SortThreeWay(arr []int) {
    qs.threeWayPartition(arr, 0, len(arr)-1)
}

File: frontend/src/components/benchmark/content/quick_sort/implementations/quick_sort.rs
use rand::Rng;

pub struct QuickSort;

impl QuickSort {
    fn partition(arr: &mut [i32], low: usize, high: usize) -> usize {
        let pivot = arr[high];
        let mut i = low;

        for j in low..high {
            if arr[j] <= pivot {
                arr.swap(i, j);
                i += 1;
            }
        }
        arr.swap(i, high);
        i
    }

    fn partition_random(arr: &mut [i32], low: usize, high: usize) -> usize {
        let mut rng = rand::thread_rng();
        let random_idx = rng.gen_range(low..=high);
        arr.swap(random_idx, high);
        Self::partition(arr, low, high)
    }

    fn quick_sort_recursive(arr: &mut [i32], low: usize, high: usize) {
        if low < high {
            let pi = Self::partition(arr, low, high);
            if pi > 0 {
                Self::quick_sort_recursive(arr, low, pi - 1);
            }
            Self::quick_sort_recursive(arr, pi + 1, high);
        }
    }

    fn quick_sort_random_recursive(arr: &mut [i32], low: usize, high: usize) {
        if low < high {
            let pi = Self::partition_random(arr, low, high);
            if pi > 0 {
                Self::quick_sort_random_recursive(arr, low, pi - 1);
            }
            Self::quick_sort_random_recursive(arr, pi + 1, high);
        }
    }

    pub fn sort(arr: &mut [i32]) {
        if !arr.is_empty() {
            Self::quick_sort_recursive(arr, 0, arr.len() - 1);
        }
    }

    pub fn sort_random(arr: &mut [i32]) {
        if !arr.is_empty() {
            Self::quick_sort_random_recursive(arr, 0, arr.len() - 1);
        }
    }

    pub fn sort_iterative(arr: &mut [i32]) {
        if arr.is_empty() {
            return;
        }

        let mut stack = vec![(0, arr.len() - 1)];

        while let Some((low, high)) = stack.pop() {
            if low < high {
                let pi = Self::partition(arr, low, high);
                if pi > 0 {
                    stack.push((low, pi - 1));
                }
                stack.push((pi + 1, high));
            }
        }
    }

    fn three_way_partition(arr: &mut [i32], low: usize, high: usize) {
        if low >= high {
            return;
        }

        let mut lt = low;
        let mut gt = high;
        let pivot = arr[low];
        let mut i = low;

        while i <= gt {
            if arr[i] < pivot {
                arr.swap(lt, i);
                lt += 1;
                i += 1;
            } else if arr[i] > pivot {
                arr.swap(i, gt);
                if gt == 0 {
                    break;
                }
                gt -= 1;
            } else {
                i += 1;
            }
        }

        if lt > 0 {
            Self::three_way_partition(arr, low, lt - 1);
        }
        Self::three_way_partition(arr, gt + 1, high);
    }

    pub fn sort_three_way(arr: &mut [i32]) {
        if !arr.is_empty() {
            Self::three_way_partition(arr, 0, arr.len() - 1);
        }
    }
}